test/e2e/agent/agent_test.go

```diff
@@ -63,4 +65,207 @@ var _ = Describe("VM Agent behavior", func() {
+			Eventually(harness.GetDeviceWithStatusSummary, TIMEOUT, POLLING).WithArguments(
+				deviceId).Should(Equal(v1alpha1.DeviceSummaryStatusType("Online")))
+		})
+
+		It("should report the correct device status when trying to upgrade to a not existing image", func() {
+			deviceId, _ := harness.EnrollAndWaitForOnlineStatus()
+
+			var newImageReference string
+			harness.UpdateDeviceWithRetries(deviceId, func(device *v1alpha1.Device) {
+				currentImage := device.Status.Os.Image
+				logrus.Infof("Current image for %s is %s", deviceId, currentImage)
+				repo, _ := parseImageReference(currentImage)
+				newImageReference = repo + ":not-existing"
+				device.Spec.Os = &v1alpha1.DeviceOSSpec{Image: newImageReference}
+				logrus.Infof("Updating %s to image %s", deviceId, device.Spec.Os.Image)
+			})
+
+			harness.WaitForDeviceContents(deviceId, "Failed to update to renderedVersion: 2. Retrying",
+				func(device *v1alpha1.Device) bool {
+					return conditionExists(device, "Updating", "True", "Retry")
+				}, "2m")
+
+			Eventually(harness.GetDeviceWithStatusSummary, TIMEOUT, POLLING).WithArguments(
+				deviceId).Should(Equal(v1alpha1.DeviceSummaryStatusType("Degraded")))
+		})
+
+		It(`should show an error when trying to update a device with
+			"a reference to a not existing git repo, and report 'Online' status`, Label("git"), func() {
+			deviceId, _ := harness.EnrollAndWaitForOnlineStatus()
+
+			harness.UpdateDeviceWithRetries(deviceId, func(device *v1alpha1.Device) {
+				logrus.Infof("Current device is %s", deviceId)
+
+				// Create ConfigProviderSpec.
+				var configProviderSpec v1alpha1.ConfigProviderSpec
+				err := configProviderSpec.FromGitConfigProviderSpec(gitConfigInvalidRepo)
+				Expect(err).ToNot(HaveOccurred())
+
+				device.Spec.Config = &[]v1alpha1.ConfigProviderSpec{configProviderSpec}
+				logrus.Infof("Updating %s with config %s", deviceId, device.Spec.Config)
+			})
+
+			// Check the http config error is detected.
+			harness.WaitForDeviceContents(deviceId, `1 invalid configuration: example-git-config-provider. Error: failed fetching
+			specified Repository definition 00000000-0000-0000-0000-000000000000/not-existing-repo:
+			resource not found`,
+				func(device *v1alpha1.Device) bool {
+					return conditionExists(device, "SpecValid", "False", "Invalid")
+				}, "2m")
+
+			// The behaviour will change after EDM-418.
+			harness.WaitForDeviceContents(deviceId, "the device is updated to renderedVersion: 1",
+				func(device *v1alpha1.Device) bool {
+					return conditionExists(device, "Updating", "False", "Updated")
+				}, "2m")
+			Eventually(harness.GetDeviceWithStatusSummary, TIMEOUT, POLLING).WithArguments(
+				deviceId).Should(Equal(v1alpha1.DeviceSummaryStatusType("Online")))
+		})
+
+		It(`should show an error when trying to update a device with a httpConfigProviderSpec
+			with invalid Path, and report 'Online' status`, func() {
+			deviceId, _ := harness.EnrollAndWaitForOnlineStatus()
```

Comment:
I am not sure I can't avoid it. I created it to unify repeated steps, but I can't add it to the BeforeEach,  since some test before are validating the statuses after of some of the operations in the function, so I think I have to leave it.